pkg/placement/placement-resolution.go

```diff
@@ -0,0 +1,368 @@
+	// duplicated in the objectIdentifierToKey map, due to the uniqueness of the Key. Therefore, whenever an object is about to
+	// be appended to an objects slice, we simply append.
+	for _, key := range resolution.objectIdentifierToKey {
+		gvr, found := gvkGvrMapper.GetGvr(key.GVK)
+		if !found {
+			return nil, fmt.Errorf("failed to get GVR for GVK %s", key.GvkKey())
+		}
+
+		// check if object is cluster-scoped or namespaced by checking namespace
+		if key.NamespacedName.Namespace == metav1.NamespaceNone {
+			resolution.handleClusterScopedObject(gvr, key, &workload, clusterScopeDownsyncObjectsMap)
+			continue
+		}
+
+		resolution.handleNamespacedObject(gvr, key, &workload, namespaceScopeDownsyncObjectsMap, nsObjectsLocationInSlice)
+	}
+
+	return &v1alpha1.PlacementDecisionSpec{
+		Workload:     workload,
+		Destinations: destinationsStringSetToDestinations(resolution.destinations),
+	}, nil
+}
+
+// handleClusterScopedObject handles a cluster-scoped object by adding it to the workload.
+// ClusterScopeDownsyncObjectsMap is used to efficiently locate the object in the workload.
+func (resolution *placementResolution) handleClusterScopedObject(gvr schema.GroupVersionResource,
+	key *util.Key,
+	workload *v1alpha1.DownsyncObjectReferences,
+	clusterScopeDownsyncObjectsMap map[schema.GroupVersionResource]*v1alpha1.ClusterScopeDownsyncObjects) {
+	// check if obj GVR already exists in map
+	if csdObjects, found := clusterScopeDownsyncObjectsMap[gvr]; found {
+		// GVR exists, append cluster-scope object
+		csdObjects.ObjectNames = append(csdObjects.ObjectNames, key.NamespacedName.Name)
+		return
+	}
+	// GVR doesn't exist, this is the first time
+	// add ClusterScopeDownsyncResource to the workload
+	workload.ClusterScope = append(workload.ClusterScope, v1alpha1.ClusterScopeDownsyncObjects{
+		GroupVersionResource: metav1.GroupVersionResource{
+			Group:    gvr.Group,
+			Version:  gvr.Version,
+			Resource: gvr.Resource,
+		},
+		ObjectNames: []string{key.NamespacedName.Name},
+	})
+
+	// retain a pointer to the added ClusterScopeDownsyncResource for efficiency
+	clusterScopeDownsyncObjectsMap[gvr] = &workload.ClusterScope[len(workload.ClusterScope)-1]
+}
+
+// handleNamespacedObject handles a namespaced object by adding it to the workload.
+// namespaceScopeDownsyncObjectsMap and nsObjectsLocationInSlice are used to efficiently locate the object in the workload.
+func (resolution *placementResolution) handleNamespacedObject(gvr schema.GroupVersionResource,
+	key *util.Key,
+	workload *v1alpha1.DownsyncObjectReferences,
+	namespaceScopeDownsyncObjectsMap map[schema.GroupVersionResource]*v1alpha1.NamespaceScopeDownsyncObjects,
+	nsObjectsLocationInSlice map[string]int) {
+	if nsdObjects, found := namespaceScopeDownsyncObjectsMap[gvr]; found {
+		// GVR mapping is found, check NS mapping
+		gvrAndNSKey := util.KeyFromGVRandNS(gvr, key.NamespacedName.Namespace) // key for ns mapping inside
```

Comment:
FYI: This key construction is needed if `!found` too. I would hoist this above the `if`.

Reply:
Job half done. The reduant call to `util.KeyFromGVRandNS` is still there.

Reply:
Missed the one that was not referred to in this comment originally, updated.